### PR TITLE
Add name parameter to to_thread.run_sync [SQUASH]

### LINKED ISSUE #1148: Optionally allow to set the thread's name where possible
This request is kind of specific to Celery's needs but I'm sure other's may agree this is worthwhile doing.

On platforms which support this behaviour (Linux, MacOS, *BSD, New versions of Windows) it is possible to set the thread's name (or description on Windows).

It requires some C/C++ or an optional dependency but shouldn't be that hard to implement.

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #1148

This doesn't work on windows, requiring an entirely separate mechanism (https://stackoverflow.com/a/16199587). It should be possible to get working on Mac (https://stackoverflow.com/a/7989973), but not having a mac machine I need the CI to test that for me. BSD variants also might need slightly different code, but I won't bother with that not having a way to test it. I haven't added a decorator on the test, so will probably see failed tests on other OS's for now.

The default in #1148 was suggested to be "Thread for {trio.lowlevel.current_task().name" by @njsmith - but pthread limits thread names to 15 characters so that's kind of a bad default. Wasn't sure what else to have though so that's the current default, but we probably want something else.

I currently set the thread name from within the thread, but it's possible to set it from the parent thread (although maybe not on mac) but the first way I found out of getting a thread id was with `threading.get_ident()` so I did it in the child.

Options for setting the name was `pthread`, `sys/prctl.h` and `libcap.so`. The first should be the one most commonly found on systems, being installed with `glibc` (on my Arch system), https://stackoverflow.com/a/51668422 says `prctl` depends on `libcap-dev` (though it's also bundled with `glibc` on mine), and `libcap.so` is for me dependent on `libcap` which itself depends on `glibc`.

Actually seeing the thread name in `ps` turned out to be flaky with everybody on SO saying different things, only `ps -eT` worked for me.

Making it a draft until it's been tested on other systems.

@Zac-HD

### CODE REVIEW

**Comment by jakkdl** [comment]
Oh and if the thread name should be possible to set (or get) in any other functions or exposed in any way that shouldn't be too hard to implement.

File: trio/_threads.py
Comment (Zac-HD) [inline trio/_threads.py:63]: I prefer `thread_name=None` here, to make it clear what the name is for, and that we'll construct a default if not given a name.

File: trio/_threads.py
Comment (Zac-HD) [inline trio/_threads.py:231]: This will be a little fiddly to plumb through, but we can do better: because Trio ultimately uses `threading.Thread` under the hood, we can assign to the `.name` attribute which should work on all platforms.  Following `start_thread_soon` through will get you to `trio/_core/_thread_cache.py`.

Additional complication: we actually have a cached pool of threads, so we'll want to set the name when we starting working, and reset to a generic e.g. `trio-idlethread` when the task is over.

Comment (jakkdl) [reply trio/_threads.py:231]: could avoid having to plumb through several levels of call by using `threading.current_thread()` instead of `.get_ident()`.

That doesn't sound to bad, just unset the name after `sync_fn(*args)` is called in `worker_fn`? Though I guess that might fit better in `_core/_thread_cache` and is an argument for doing proper plumbing and placing all the logic in there.

**Comment by jakkdl** [comment]
Pushed what a solution might look like, but I'm quite unhappy with it - ugly, cumbersome and probably slows down thread operations a bit even if you're not using names (since it will try to find the pthread library anew in each new thread). And given that `_thread_cache` talks about memory cache behavior I presume the performance here is somewhat important. Working on moving it into `_thread_cache` and speeding up the default case of not naming your threads.

Though there's a question there of if you should bother re-naming idle and active threads (probably not?) as well as how worth it is keeping the name counter in the name (means the thread needs to remember the name/number when renaming so it can restore it)

File: trio/_threads.py
Comment (Zac-HD) [reply trio/_threads.py:236]: Yeah, I'd definitely put it in https://github.com/python-trio/trio/blob/ce4c99e696a516db40b55fb076deca2e586626cf/trio/_core/_thread_cache.py#L59-L70

Then we can just do `self.thread.name = thread_name` before `result = outcome.capture(fn)` and reset afterwards.  It also doesn't look like this interface shares the 15-character limit?

Comment (jakkdl) [reply trio/_threads.py:236]: yeah the 15-char limit is just in pthreads / the OS, thread.name is only used internally in python so that's a normal unicode string.

**Comment by jakkdl** [comment]
Okay this is starting to look like a decent solution. Question remains on how to handle naming idle threads, and default names in general. But I think otherwise it's in a decent spot.

Would it be possible to give me access as if I'm not a first-time contributor so I can iterate on getting it to work / not crash horribly on other OS's? Gonna take ages to finish this if I have to wait for approval on each workflow run.

**Comment by jakkdl** [comment]
Okay this should be pretty decent.

`ctypes.util.find_library("pthread")` fails on the CI mac machines, which I didn't expect, so I'm going to leave getting it to work on mac to somebody else.

**Review by Zac-HD — COMMENTED** [review]
With a few comments below, this looks good to me - thanks @jakkdl!

I'm also unsure whether changing OS thread names per-task is going to be a net gain for debuggability; I lean yes but will want to check in with some colleagues and @oremanj before we merge.

File: trio/_core/_thread_cache.py
Comment (jakkdl) [reply trio/_core/_thread_cache.py:39]: Oops, yep!

File: trio/_threads.py
Comment (jakkdl) [reply trio/_threads.py:184]: Yeah I'm not sure how to default it, 15 chars really isn't a lot and can easily get hit just with the function name. Could do `f'{trio.lowlevel.current_task().name.split(".")[-1]'[:14-len("thread_number") {thread number}]`

I think the last part of the name is usually the most interesting part? Or maybe `.__name__` on the passed function. And I think you might want the thread number in case you're launching several identical ones and want to disambiguate them.

**Comment by jakkdl** [comment]
It seems to work fully on Mac now! The macOS 3.7 doesn't have pthreads apparently, but the later ones do and they pass the tests with flying colors. And windows doesn't crash (and sets Thread.name).
The codecov fail is due to differing platform behaviour, and one weird edge case of pthread but no pthread_setname_np that def should be pragma'd. But I'm not sure what the praxis is on the other platform stuff, I can litter everything non-linux with `#pragma: no cover` if you want.

**Review by Fuyukai — APPROVED** [review]
looks fine to me, idrc about the default thread name that much as there's no easy way to decide it. it's up to you if you want to change it or not.

File: trio/_threads.py
Comment (Zac-HD) [reply trio/_threads.py:182]: I find the thread-ID point pretty convincing, so how about we just `thread_name.removeprefix("Thread for ")` before truncating and assigning the OS thread name?  See other comment for details.

File: trio/_core/_thread_cache.py
Comment (Zac-HD) [inline trio/_core/_thread_cache.py:15]: Let's add a helper function:

```python
def _to_os_thread_name(name: str) -> bytes:
    if name.startswith("Thread for "):  # str.removeprefix is new in 3.9
        name = name.replace("Thread for ", "")
    return name.encode("ascii", errors="replace")[:16]  # 15 chars + implicit trailing \x00
```

File: trio/_core/_thread_cache.py
Comment (Zac-HD) [inline trio/_core/_thread_cache.py:21]: You comment below that `on mac OSX pthread_setname_np does not take a thread id`, but the function does actually take that - bug?  If they *can* have the same function signatures then there's some duplicate code instead.

**Review by Zac-HD — COMMENTED** [review]
The coverage looks fine to me.  See comments below; after that I think we're ready to merge!

File: trio/_core/_thread_cache.py
Comment (jakkdl) [reply trio/_core/_thread_cache.py:21]: The code calling namefunc will always pass a thread id, so it doesn't need to differ between platforms. But `setname` doesn't get `ident` passed to it.
I wasn't sure what to do if `ident is None` in `darwin_namefunc`. Documentation says that happens if the thread isn't started, but I don't really know what that means wrt naming the thread, but I went with the safe option of not doing anything in that case. (and idk if it *can* happen given Trio's usage of it)
Will add some comment to it.

File: trio/_threads.py
Comment (jakkdl) [reply trio/_threads.py:182]: Oh actually, I think removing the prefix could trip up the end user. Somebody might manually assign "Thread for [a/b/c/d/e]" as their thread name. I'm ditching "Thread for" entirely, will push up a new version soon.

File: trio/_core/_thread_cache.py
Comment (jakkdl) [reply trio/_core/_thread_cache.py:15]: btw `[:16]` doesn't work - it seems ctypes adds the null byte on it's own so it fails on anything more than `[:15]`

**Comment by jakkdl** [comment]
Ubuntu3.7 seems to consistently fail for not-my-fault reasons.
I pragma'd all branches which I don't think ever gets entered, but codecov doesn't seem to include the macOS CI runs so 57-58 is the only failing part atm.

**Comment by jakkdl** [comment]
So yeah afaik this should be mergeable now! @Zac-HD

File: trio/tests/test_threads.py
Comment (Zac-HD) [inline trio/tests/test_threads.py:214]: Uncomment or delete?

**Review by Zac-HD — APPROVED** [review]
The code itself looks good to me, but I'm not comfortable merging while we have CI failing with an SSL-related error.

File: trio/tests/test_threads.py
Comment (jakkdl) [reply trio/tests/test_threads.py:214]: I guess line spacing didn't make it clear, but the comment on line 210 refers to 211 vs 212-214

**Comment by jakkdl** [comment]
The SSL errors looks to be unrelated #2488

File: trio/tests/test_threads.py
Comment (jakkdl) [reply trio/tests/test_threads.py]: but probably unnecessary, updated the comment and removed the commented out lines.

**Comment by Zac-HD** [comment]
🎉 Thanks again to everyone involved in this (un?) surprisingly long pull request!

**Pull request merged**
